src/Microsoft.ApiDesignGuidelines.Analyzers/Core/PassSystemUriObjectsInsteadOfStrings.cs

```diff
@@ -24,15 +29,202 @@ public abstract class PassSystemUriObjectsInsteadOfStringsAnalyzer : DiagnosticA
+                if (!stringParameters.Any())
+                {
+                    // no string parameter. not interested.
+                    return false;
+                }
+
+                // now do cheap string check whether those string parameter contains uri word list we are looking for.
+                if (!CheckStringParametersQuick(stringParameters, cancellationToken))
+                {
+                    // no string parameter that contains what we are looking for.
+                    return false;
+                }
+
+                // now we make sure we actually have overloads that contains uri type parameter
+                if (!CheckOverloads(model, node, cancellationToken))
+                {
+                    // no overload that contains uri as parameter
+                    return false;
+                }
+
+                // now we do more expensive word parsing to see whether this is really right one to report it.
+                return CheckStringParametersSlow(stringParameters, cancellationToken);
+            }
+
+            private bool CheckOverloads(SemanticModel model, SyntaxNode node, CancellationToken cancellationToken)
+            {
+                var uriType = _uri;
+                foreach (var overload in model.GetMemberGroup(node, cancellationToken).OfType<IMethodSymbol>())
+                {
+                    cancellationToken.ThrowIfCancellationRequested();
+
+                    if (overload.Parameters.Any(p => p.Type?.Equals(uriType) == true))
+                    {
+                        return true;
+                    }
+                }
+
+                return false;
+            }
+
+            private bool CheckStringParametersQuick(IEnumerable<IParameterSymbol> stringParameters, CancellationToken cancellationToken)
+            {
+                foreach (var parameter in stringParameters)
+                {
+                    if (CheckStringParameter(parameter, cancellationToken))
+                    {
+                        return true;
+                    }
+                }
+
+                return false;
+            }
+
+            private static bool CheckStringParameter(IParameterSymbol parameter, CancellationToken cancellationToken)
+            {
+                foreach (var word in uriWords)
+                {
+                    cancellationToken.ThrowIfCancellationRequested();
+
+                    if (parameter.Name?.IndexOf(word, StringComparison.OrdinalIgnoreCase) >= 0)
+                    {
+                        return true;
+                    }
+                }
+
+                return false;
+            }
+
+            private bool CheckStringParametersSlow(IEnumerable<IParameterSymbol> stringParameters, CancellationToken cancellationToken)
+            {
+                foreach (var parameter in stringParameters)
+                {
+                    if (parameter.Name == null ||
```

Comment:
not sure whether that is true. and I don't believe null check here cause anything to slow down so I will just leave in paranoid side.